### ISSUE TITLE
Incorrect Nonlinearity Reference File for NIRCam

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#7067]
_Issue [JP-2537](https://jira.stsci.edu/browse/JP-2537) was created on JIRA by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan):_

During OTE commissioning I noticed that the outer portions of the PSFs of every Primary Mirror Segment Assembly (PMSA) PSF there were clusters of pixels that were flagged with the JUMP DQ bit. As I looked into the post-nonlinearity correction images I found that the difference between the last groups and the second to the last group was being flag as a jump because it was more than five sigma above the median difference. This large difference is due to an over correction in the non-linearity correction.

Attached is an image of a region of PMSA PSFs (array.jpg). The second image is the DQ extension of the rate.fits image. The dark red pixels are the ones correctly flagged as SATURATED. The orange pixels are the ones flagged as JUMP. The jumps are all in the last sample. 

Note that the white pixels in the DQ image are ones that are flagged as NO SAT CHECK. This is due to the problem raised in JP-2525. The combination of these two effects is significantly affecting the shape of the PSF.

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576132#comment-576132):_

NIRCam will have to look into this. After discussion with [Karl Misselt](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kmisselt), two possibilities are:

1) Differences in the bias level at the current elevated operating temperature, which, after superbias subtraction, would push the pixel value onto an incorrect location on the non-linearity curve.
2) Our so-called "weird" pixels, which receive leaked charge when a neighboring pixel saturates. <https://outerspace.stsci.edu/display/JWSTCC/Weird+Pixels>

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576276#comment-576276):_

It's possible that the bias level is not correct. This is one of the weaknesses of the pipeline's superbias step.  It's not clear to me how the interaction between the superbias and the saturation and nonlinearity all play well together. The superbias subtraction is not as simple as I thought and is connected to the reference pixel correction. All of these steps need to collaborate for the jump step to work correctly.  

The incorrect jump flagging is saying that line is curved.

**Comment by stscijgbot-jp** [comment]
_Comment by [Karl Misselt](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kmisselt) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576338#comment-576338):_

Re: DQ mask.  Those pixels should not be NO_SAT_CHECK.  Those are the ones with a deeper well and hence get leaked charge from neighbor saturated pixels when flood illuminated as in ground test.  They do have a saturation level determined.  They are set in the well depth file to the point at which their 'response changes', i.e. they start getting charge from neighboring pixels.  So their empirical well depth is effectively the well depth of their neighbors, but offset downward because they start a lower level  They were tagged in the DQ mask to indicate that the well depth was not the true well depth, but  they should NOT be tagged so as not to have a saturation check.  So if we are actually tagging these to not find saturation, that's a bug in the pipeline.  They should have a valid well albeit significantly lower than their true well.

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576349#comment-576349):_

[Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) what program and observation are the images you show above?

**Comment by stscijgbot-jp** [comment]
_Comment by [Karl Misselt](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kmisselt) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576351#comment-576351):_

Re: superbias, reference pixels, etc.  Yes, the superbias depends on the reference pixel correction being applied (which is why we shouldn't cut out superbias from full frame for subarrays, but rather derive independent bias frames - we will have the data for that). Speaking from what I did for test data - I'm not sure how all this is implemented in the pipeline - the bias frames were constructed from reference corrected darks.  They represent the residual pedestal not removed by the reference correction which is of the order of several thousand DN.   Th
e linearity correction was derived wrt to the reference subtracted, bias subtracted pedestal.  So one either has to do those steps prior to linearity correction or empirically reference back to ~0 pedestal in the linearity correction.

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576414#comment-576414):_

Nevermind, I was looping over the wrong set of mnemonics. I see the NO_SAT_CHECKS now.
-Also, [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan), can you tell me which mask reference file and saturation reference file were used for your data? I just checked what I believe are the current mask and saturation reference files (I had to guess on the mask files since the CRDS server is currently down), and I did not find any pixels flagged as NO_SAT_CHECK in any of them. I want to make sure I'm looking in the right place.-

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576983#comment-576983):_

I've created a set of new saturation reference files with better mappings of the DQ mnemonics from Karl's original to DMS. There are no more NO_SAT_CHECK pixels. However, this doesn't solve the problem of the pixels in the PSF wings being flagged as jumps. Looking at the signal in one of these pixels before and after linearity correction, it appears that the jump flags are appropriate:

This if for NRC-05 obs 4 BLONG data, jw01059004001_02103_00001_nrcblong_uncal.fits, pixel (x,y) = (1428, 527)
saturation level: 53642.7

uncal file signal:                16388, 19232, 22487, 30566, 45756
jump step output signal:   2606.5286,  5544.1   ,  8997.292 , 17822.56  , 35699.668

CDS signals in uncal: 2844,  3255,  8079, 15190
CDS signals in jump:  2937.5715,  3453.192 ,  8825.269 , 17877.107

DQ flags in jump step output: 0, 0, 4, 4, 4   (4 is the bit for a detected jump)

Perhaps this comes down to our "weird" pixels? In this case, the pixel is in the wings of the PSF. Several pixels in the center of the PSF are saturated all the way up the ramp. So perhaps signal from these saturated pixels is leaking into the neighboring pixels? And with more pixels saturating as the ramp progresses, this leads to more leaked signal in the PSF wings, which is why the signal RATE is increasing with time??

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576988#comment-576988):_

The pixel immediately below this pixel saturates in group 4 of the ramp, which is consistent with the signal rate in the initial pixel increasing from that point onward.

**Comment by stscijgbot-jp** [comment]
_Comment by [Karl Misselt](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kmisselt) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=577013#comment-577013):_

I think this (1428,527) is result of a charge cascade on a saturating PSF.  Very simplistically (i.e. a bunch of other pixels are saturating, so it's not a simple linear tree like this) with reference to the attached plot, pixel [1428,525] (black curve) saturates on the second read, leaking charge into [1428,526] and changing its response between reads 2 and 3.  [1428,526] is thereby 'accelerated' towards saturation and fully saturates by read 4.  [1428,526] (blue curve) subsequently is leaking charge to [1428,527] (green curve) changing its response.  [1428,527] actually changes response between group 3 and 4, so there are probably more saturating neighbor pixels affecting it.  Ideally, in this situation, we should end up with the following scenario

*  [1428,525] is tagged as saturating on group 2, unusable
*  [1428,526] is tagged as 'saturating'/charge leak compromised  on group 3, use only groups 1-2 in ramp fitting. 
*  [1428,527] is tagged as 'saturating'/charge leak compromised on group 4, use only groups 1-3 in ramp fitting 

The upshot is that, we cannot rely on a static saturation map, but rather need a dynamic one. Roughly when a pixel saturates, it's neighbors must be defined as compromised on the same group and any subsequent processing (jump detection, ramp fitting) occurs only on groups before the compromise.  This has to be a linked list of sorts, since even the  compromised pixels need to continue to be tracked so when they reach true saturation and start leaking themselves, we tag their neighbors as 'compromised' at that group and continue the process.  Essentially we can't simple read in a saturation mask and label a pixel  as saturated.  We need to read a mask, find all pixels that saturate and sort their neighbors.  Should be relatively straightforward (he said) and before we start any processing, we've created a dynamic saturation/compromised map which tells the rest of the pipeline which groups are actually usable for each pixel.

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576132#comment-576132):_

NIRCam will have to look into this. After discussion with [Karl Misselt](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kmisselt), two possibilities are:

1) Differences in the bias level at the current elevated operating temperature, which, after superbias subtraction, would push the pixel value onto an incorrect location on the non-linearity curve.
2) Our so-called "weird" pixels, which receive leaked charge when a neighboring pixel saturates. <https://outerspace.stsci.edu/display/JWSTCC/Weird+Pixels>

**Comment by stscijgbot-jp** [comment]
_Comment by [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576276#comment-576276):_

It's possible that the bias level is not correct. This is one of the weaknesses of the pipeline's superbias step.  It's not clear to me how the interaction between the superbias and the saturation and nonlinearity all play well together. The superbias subtraction is not as simple as I thought and is connected to the reference pixel correction. All of these steps need to collaborate for the jump step to work correctly.  

The incorrect jump flagging is saying that line is curved.

**Comment by stscijgbot-jp** [comment]
_Comment by [Karl Misselt](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kmisselt) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576338#comment-576338):_

Re: DQ mask.  Those pixels should not be NO_SAT_CHECK.  Those are the ones with a deeper well and hence get leaked charge from neighbor saturated pixels when flood illuminated as in ground test.  They do have a saturation level determined.  They are set in the well depth file to the point at which their 'response changes', i.e. they start getting charge from neighboring pixels.  So their empirical well depth is effectively the well depth of their neighbors, but offset downward because they start a lower level  They were tagged in the DQ mask to indicate that the well depth was not the true well depth, but  they should NOT be tagged so as not to have a saturation check.  So if we are actually tagging these to not find saturation, that's a bug in the pipeline.  They should have a valid well albeit significantly lower than their true well.

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576349#comment-576349):_

[Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan) what program and observation are the images you show above?

**Comment by stscijgbot-jp** [comment]
_Comment by [Karl Misselt](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kmisselt) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576351#comment-576351):_

Re: superbias, reference pixels, etc.  Yes, the superbias depends on the reference pixel correction being applied (which is why we shouldn't cut out superbias from full frame for subarrays, but rather derive independent bias frames - we will have the data for that). Speaking from what I did for test data - I'm not sure how all this is implemented in the pipeline - the bias frames were constructed from reference corrected darks.  They represent the residual pedestal not removed by the reference correction which is of the order of several thousand DN.   Th
e linearity correction was derived wrt to the reference subtracted, bias subtracted pedestal.  So one either has to do those steps prior to linearity correction or empirically reference back to ~0 pedestal in the linearity correction.

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576414#comment-576414):_

Nevermind, I was looping over the wrong set of mnemonics. I see the NO_SAT_CHECKS now.
-Also, [Michael Regan](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mregan), can you tell me which mask reference file and saturation reference file were used for your data? I just checked what I believe are the current mask and saturation reference files (I had to guess on the mask files since the CRDS server is currently down), and I did not find any pixels flagged as NO_SAT_CHECK in any of them. I want to make sure I'm looking in the right place.-

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576983#comment-576983):_

I've created a set of new saturation reference files with better mappings of the DQ mnemonics from Karl's original to DMS. There are no more NO_SAT_CHECK pixels. However, this doesn't solve the problem of the pixels in the PSF wings being flagged as jumps. Looking at the signal in one of these pixels before and after linearity correction, it appears that the jump flags are appropriate:

This if for NRC-05 obs 4 BLONG data, jw01059004001_02103_00001_nrcblong_uncal.fits, pixel (x,y) = (1428, 527)
saturation level: 53642.7

uncal file signal:                16388, 19232, 22487, 30566, 45756
jump step output signal:   2606.5286,  5544.1   ,  8997.292 , 17822.56  , 35699.668

CDS signals in uncal: 2844,  3255,  8079, 15190
CDS signals in jump:  2937.5715,  3453.192 ,  8825.269 , 17877.107

DQ flags in jump step output: 0, 0, 4, 4, 4   (4 is the bit for a detected jump)

Perhaps this comes down to our "weird" pixels? In this case, the pixel is in the wings of the PSF. Several pixels in the center of the PSF are saturated all the way up the ramp. So perhaps signal from these saturated pixels is leaking into the neighboring pixels? And with more pixels saturating as the ramp progresses, this leads to more leaked signal in the PSF wings, which is why the signal RATE is increasing with time??

**Comment by stscijgbot-jp** [comment]
_Comment by [Bryan Hilbert](https://jira.stsci.edu/secure/ViewProfile.jspa?name=hilbert) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=576988#comment-576988):_

The pixel immediately below this pixel saturates in group 4 of the ramp, which is consistent with the signal rate in the initial pixel increasing from that point onward.

**Comment by stscijgbot-jp** [comment]
_Comment by [Karl Misselt](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kmisselt) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=577013#comment-577013):_

I think this (1428,527) is result of a charge cascade on a saturating PSF.  Very simplistically (i.e. a bunch of other pixels are saturating, so it's not a simple linear tree like this) with reference to the attached plot, pixel [1428,525] (black curve) saturates on the second read, leaking charge into [1428,526] and changing its response between reads 2 and 3.  [1428,526] is thereby 'accelerated' towards saturation and fully saturates by read 4.  [1428,526] (blue curve) subsequently is leaking charge to [1428,527] (green curve) changing its response.  [1428,527] actually changes response between group 3 and 4, so there are probably more saturating neighbor pixels affecting it.  Ideally, in this situation, we should end up with the following scenario

*  [1428,525] is tagged as saturating on group 2, unusable
*  [1428,526] is tagged as 'saturating'/charge leak compromised  on group 3, use only groups 1-2 in ramp fitting. 
*  [1428,527] is tagged as 'saturating'/charge leak compromised on group 4, use only groups 1-3 in ramp fitting 

The upshot is that, we cannot rely on a static saturation map, but rather need a dynamic one. Roughly when a pixel saturates, it's neighbors must be defined as compromised on the same group and any subsequent processing (jump detection, ramp fitting) occurs only on groups before the compromise.  This has to be a linked list of sorts, since even the  compromised pixels need to continue to be tracked so when they reach true saturation and start leaking themselves, we tag their neighbors as 'compromised' at that group and continue the process.  Essentially we can't simple read in a saturation mask and label a pixel  as saturated.  We need to read a mask, find all pixels that saturate and sort their neighbors.  Should be relatively straightforward (he said) and before we start any processing, we've created a dynamic saturation/compromised map which tells the rest of the pipeline which groups are actually usable for each pixel.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-2537?focusedCommentId=938269#comment-938269):_

Digging through old open tickets: I think this one is no longer relevant given changes made to the pipeline and reference files in the last few years.  Specifically, the pipeline now applies multiple passes to the saturation check, expanding flagging to neighbors of saturated pixels when they pass the relevant threshold.  Closing.
